rfd/0000-sre-challenge.md

```diff
@@ -0,0 +1,264 @@
+### Design Document for Kubernetes Deployment Scaler
+
+---
+authors: Derek Norrbom (dereknorrbom@gmail.com)
+state: draft
+---
+
+## What
+
+This document outlines the proposed design and implementation plan for a Kubernetes Deployment Scaler API. The API will provide functionality to get and set the replica count of a Kubernetes deployment, utilizing mTLS for secure communication and caching for efficient data retrieval.
+
+## Why
+
+The purpose of this project is to implement a scalable, secure API for managing Kubernetes deployments. This project will demonstrate proficiency in Go, Kubernetes, Docker, and security best practices (mTLS). It will also highlight the ability to design, build, and document an application that can be easily extended and maintained.
+
+## How
+
+### API Structure
+
+The API will provide the following endpoints:
+
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running and Kubernetes connectivity is verified.
+- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
+- **POST /set-replica-count?namespace={namespace}&deployment={deployment}**: Sets the replica count of the specified deployment.
+- **GET /deployments?namespace={namespace}**: Lists all deployments in the specified namespace.
+
+Each endpoint will be secured using mutual TLS (mTLS) to ensure secure communication between the client and the server.
+
+#### Example JSON Responses
+
+- **GET /healthz**
+   ##### Response:
+   ```json
+   {
+      "status": "OK"
+   }
+   ```
+
+- **GET /replica-count?namespace=default&deployment=my-deployment**
+   ##### Response:
+   ```json
+   {
+      "replicaCount": 3
+   }
+   ```
+
+- **POST /set-replica-count?namespace=default&deployment=my-deployment**
+   ##### Request Body
+   ```json
+   {
+      "replicas": 3
+   }
+   ```
+   ##### Response
+   ```json
+   {
+      "message": "Replica count updated to: 3"
+   }
+   ```
```

Comment:
The response message reads a bit like a log message. What do you think about returning the same message as returned from the get replica count endpoint?

Reply:
Returning the same message as the GET replica count endpoint makes sense for consistency. I'll update the response to match the GET endpoint.